Cart and Checkout block transforms for classic shortcodes

mikejolley opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

This PR improves the incompatibility flow for cart and checkout blocks when non compatible extensions are detected. 

**Task List**

Adds a task list item to review the cart/checkout blocks. This notice is only visible if using blocks on the cart/checkout pages, and will redirect to either the checkout page, or cart page, on click.

![Screenshot 2023-10-17 at 14 51 37](https://github.com/woocommerce/woocommerce-blocks/assets/90977/647b2bfc-52d5-4813-82e6-39ee11389471)

**Incompatibility notices**

The notice wording is updated, and a convenient switch button is introduced to revert to classic shortcodes:

![Screenshot 2023-10-17 at 14 55 03](https://github.com/woocommerce/woocommerce-blocks/assets/90977/ffc9d426-be11-4529-b0a0-a2542d103189)

Clicking the switch button opens a confirmation model to make sure the merchant understands the ramifications of reverting to shortcodes:

![Screenshot 2023-10-17 at 14 56 33](https://github.com/woocommerce/woocommerce-blocks/assets/90977/b3afd06e-8552-438c-abf3-28d36793d274)

Confirming will replace the blocks with the classic shortcode. There is a snackbar shown giving the option to undo the action:

![Screenshot 2023-10-17 at 14 57 15](https://github.com/woocommerce/woocommerce-blocks/assets/90977/c471ee41-b7ba-4bcd-ac8f-79ff5711e1df)

**Block Transforms**

This PR also includes transforms to convert from block->shortcode even if the notice is hidden:

![Screenshot 2023-10-17 at 15 15 45](https://github.com/woocommerce/woocommerce-blocks/assets/90977/5acbb62b-e7ce-4b48-bd07-30d3ba32dba3)

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/10848

## Why

This allows easier switching back to shortcodes if the block based checkout is not working or suitable for a store, and encourages merchants to check their pages are correctly configured.

## Follow ups

cc @ralucaStan Creating some follow ups so we can at least ship something in time in case these turn out to be complex tasks.

- [x] https://github.com/woocommerce/woocommerce-blocks/issues/11314
- [x] Find a way to programmatically open the block inspector based on a query string variable when navigating from the task list https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/587154835a778da20dfa35261703b4785d3b3baf

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

Prerequisite: Ensure you're already using cart and checkout blocks and you have an incompatible extension installed. You can use `WooCommerce Checkout Field Editor` as an example.

1. Go to WooCommece > Home. There should be a task list item "Review your shopper's checkout experience". Click it. You should be taken to the page editor.
2. Edit the cart page
    1. Edit the cart page
    2. Focus on the parent cart block and click its icon. You should be able to transform to "classic cart"
    3. Refresh the editor
    4. Focus on the cart block. See the incompatibility notice in the sidebar. Click the "switch to classic cart" button.
    5. Modal should open. Click switch. Block should be replaced.
    6. In the snackbar bottom left, click "undo". Should revert back to blocks.
3. Repeat above tests for **checkout block**. Pay attention to wording and ensure any references to "cart" are updated with "checkout" during the process.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

https://github.com/woocommerce/woocommerce-blocks/assets/90977/c923f771-cf6e-4905-8440-a70a9d2679bc

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Added tools to switch between Cart and Checkout Blocks and Classic Shortcodes when non-compatible payment extensions are detected.

mikejolley committed:
Add transforms for checkout block

mikejolley committed:
Block to shortcode switcher in notice

mikejolley committed:
cart transforms

mikejolley committed:
Fix target block for switching

mikejolley committed:
Remove switcher UI for classic cart/checkout

mikejolley committed:
Set isPreview when generating block preview in switcher

mikejolley committed:
Onboarding task

mikejolley committed:
Action on click

mikejolley committed:
Focus on block after replacement

mikejolley committed:
Update notice styling and wording

mikejolley committed:
Undo functionality

mikejolley committed:
Look for woocommerce/classic-shortcode when determining if task list item should display

senadir commented on assets/js/blocks/classic-shortcode/checkout.tsx:
```diff
@@ -50,7 +50,7 @@ const onClickCallback = ( {
 const getTitle = () => {
-	return __( 'Checkout Shortcode', 'woo-gutenberg-products-block' );
+	return __( 'Checkout Checkout', 'woo-gutenberg-products-block' );
 };
```
Checkout checkout?

mikejolley replied:
Start spreadin' the news. I'm leavin' today. I want to be a part of it. 

Checkout Checkout

senadir replied:
🤣

senadir commented on assets/js/blocks/classic-shortcode/index.tsx:
```diff
@@ -56,31 +56,44 @@ const conversionConfig: { [ key: string ]: BlockifiedTemplateConfig } = {
 
-		return [ ...acc, block.clientId ];
-	}, [] );
+	if ( groupBlock ) {
+		return (
+			groupBlock.innerBlocks.find( ( block ) => {
+				return (
+					block.name === 'woocommerce/cart' ||
+					block.name === 'woocommerce/checkout'
+				);
+			} )?.clientId || ''
+		);
+	}
+};
```
Isn't this logic already covered by the groupBlock logic?

mikejolley replied:
Both loops are needed; first to find the parent group block, and 2nd to select the client ID of the cart/checkout block within it.

I've improved the code slightly in https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/b16cd4c131b537a3f6eee6df7f54da837bdc795a to support cases where a group is not present.

senadir commented on assets/js/editor-components/incompatible-extension-notice/index.tsx:
```diff
@@ -36,15 +60,26 @@ export function IncompatibleExtensionsNotice( {
+					'Switch to classic checkout',
+					'woo-gutenberg-products-block'
+			  );
+	const blockLabel = block === 'woocommerce/cart' ? 'cart' : 'checkout';
```
string is not translated here.

senadir commented on assets/js/editor-components/incompatible-extension-notice/index.tsx:
```diff
@@ -36,15 +60,26 @@ export function IncompatibleExtensionsNotice( {
-						__(
-							'The following extensions may be incompatible with the block-based checkout. <a>Learn more</a>',
-							'woo-gutenberg-products-block'
+						sprintf(
+							// translators: %s is the name of the parent block.
+							__(
+								'Some extensions do not yet support the new %s and may impact the shopper experience. <a>Learn more</a>',
+								'woo-gutenberg-products-block'
+							),
+							blockLabel
```
This is a nitpick but ideally you just do this string twice (above) without having cart/checkout as placeholder.

Cart is a single word in arabic, checkout is 2 words, depending on what comes before it, it might be translated with an article or not, but having it separately wouldn't make that obvious.

mikejolley replied:
https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/45d9543abe479064e7d9d87ad2180025852cc5a2

senadir commented on src/Domain/Bootstrap.php:
```diff
@@ -129,6 +130,7 @@ function() {
 		$this->container->get( DraftOrders::class )->init();
 		$this->container->get( CreateAccount::class )->init();
 		$this->container->get( ShippingController::class )->init();
+		$this->container->get( TasksController::class )->init();
```
This can probably be below inside the `! $is_rest`

mikejolley replied:
Not in this case. I thought the same, but the task list uses the API to get active tasks. Moving it prevents us displaying anything.

senadir reviewed (approved):
This is testing fine for me and while most of the code was new to me, I did manage to understand and leave some feedback.

Thank you for working on this Mike!

mikejolley committed:
Enable focus on the cart/checkout block when visiting from the task list

mikejolley commented:
Thanks @senadir Ill work on that feedback. Also note this commit for one of the follow up tasks https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/587154835a778da20dfa35261703b4785d3b3baf

mikejolley committed:
Classic Cart/Checkout Updated Title

mikejolley committed:
Add missing translations

pmcpinto commented:
@mikejolley I've tested the flow and it works well. Great job. I only have comments about some details:
- In this p2 comment (pdFofs-1qA-p2#comment-1866), I've suggested redirecting to the checkout when the store uses both blocks. The flow that you implemented redirects the user to the Cart. After testing it, this change makes sense, as it follows the shopper flow: first the cart and then the checkout. @jarekmorawski do you agree?
- Can we add telemetry to the editor? Namely, the "Switch to classic X" action in the notification and the actions available in the modal. In the task list, we inherit existing Tracks events (task view and task list), but in the editor we don't have anything. It would be useful to add this telemetry to have a complete picture of what's happening (view task > click task > click to switch to shortcode > confirm action)

mikejolley committed:
Refactor modal content to avoid sprintf

ralucaStan commented:
@mikejolley I see in your video you tested with a block theme. I am testing with Storefront and clicking the task link sends me to [this page](https://ccblocks.local/wp-admin/site-editor.php?postType=page&postId=59&focus=cart&canvas=edit) where I see: 
<img width="853" alt="Screenshot 2023-10-18 at 14 59 10" src="https://github.com/woocommerce/woocommerce-blocks/assets/1628454/c73c020b-289e-4c1c-aa95-0e795d4b4df8">

mikejolley committed:
Improve pickBlockClientId

ralucaStan commented:
Another small issue I found is that the undo button from the Snack Bar does not work.

<img width="825" alt="Screenshot 2023-10-18 at 15 06 37" src="https://github.com/woocommerce/woocommerce-blocks/assets/1628454/5c5d4596-af92-4f18-b99e-a3bf94f3f096">
This appears after clicking the button to transform from the classic shortcode placeholder.

mikejolley commented:
@ralucaStan can you retest undo with latest code? I submitted https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/b16cd4c131b537a3f6eee6df7f54da837bdc795a — it doesn't undo if there is no outer group block but this should be fixed now if its the same issue.

ralucaStan commented:
> @ralucaStan can you retest undo with latest code? I submitted [b16cd4c](https://github.com/woocommerce/woocommerce-blocks/commit/b16cd4c131b537a3f6eee6df7f54da837bdc795a) — it doesn't undo if there is no outer group block but this should be fixed now if its the same issue.

confirming this was fixed when using the latest build

ralucaStan commented:
I found a similar width bug with TT1 theme, but for the Cart shortcode instead. 

When I replace with the shortcode the cart page look crowded. It should be a wider. 

## current behaviour

<img width="654" alt="Screenshot 2023-10-18 at 15 27 26" src="https://github.com/woocommerce/woocommerce-blocks/assets/1628454/8a1ae178-d4d4-4a81-b60d-d61c519e04d4">

## expected result
<img width="747" alt="Screenshot 2023-10-18 at 15 30 41" src="https://github.com/woocommerce/woocommerce-blocks/assets/1628454/1f2e6be0-7ceb-4284-9b39-af798bff8174">

ralucaStan commented:
@mikejolley I updated the 2nd photo of https://github.com/woocommerce/woocommerce-blocks/pull/11228#issuecomment-1768350191 , initially I added the cart block instead

mikejolley committed:
Tracks events for switching to classic shortcode block

ralucaStan commented:
I've tested the experience with a keyboard. 
Some things I spotted:


- the buttons inside the shortcode placeholder can't be reached by keyboard
- Nice to have: after the switch to shortcode focus the shortcode placeholder. For example the block is focused after the switch

mikejolley committed:
TaskList support for non-block themes

mikejolley committed:
Updated placeholder to work on non-white page backgrounds

mikejolley committed:
Find blocks using findBlock utility

mikejolley committed:
Add TabbableContainer for buttons

mikejolley committed:
Add align to wrapper

mikejolley commented:
> I see in your video you tested with a block theme.

https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/bdf0de6ba9f1889a18f90acec9a942c131a53a31

> When I replace with the shortcode the cart page look crowded. It should be a wider.

Seems to only affect TT1 non blocks.

Fixed by adding a wrapper class: https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/a43da7dc6cf6628bbd6de661900acf9e83845dc7

> the buttons inside the shortcode placeholder can't be reached by keyboard

I added a tabbable container; you can use arrow keys to navigate to the button area and tab to switch between the buttons. Not sure if this is the optimal solution but how tabbing between items in the editor works is not clear.

Ref: https://developer.wordpress.org/block-editor/reference-guides/components/navigable-container/

> Nice to have: after the switch to shortcode focus the shortcode placeholder. For example the block is focused after the switch

Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/6d8673e7efc06d2f18f82ed465c08940d728ba80

@ralucaStan 

> Can we add telemetry to the editor? Namely, the "Switch to classic X" action in the notification and the actions available in the modal. In the task list, we inherit existing Tracks events (task view and task list), but in the editor we don't have anything. It would be useful to add this telemetry to have a complete picture of what's happening (view task > click task > click to switch to shortcode > confirm action)

Untested at present but I added 3 events in https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/659926a28de9457b335072b3aaccdfbffbf7ad43:

- switch_to_classic_shortcode_click
- switch_to_classic_shortcode_confirm
- switch_to_classic_shortcode_cancel
- switch_to_classic_shortcode_undo

@pmcpinto 

Other changes:

- Found an issue when testing TT1 where the placeholder was not visible due to the colors. Switched to SVG to fix: https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/bb1cecd499c5a7892429b30f98c580c4b3c7b28b
- Found an issue finding blocks for replacement/undo, so repurposed an existing findBlocks utility https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/6d8673e7efc06d2f18f82ed465c08940d728ba80

jarekmorawski commented:
> After testing it, this change makes sense, as it follows the shopper flow: first the cart and then the checkout. @jarekmorawski do you agree?

Yes! Do we have some logic in place to take the user to one or the other if either cart or checkout is not present? I want to make sure we're accommodating use cases where the user only has block checkout, but not cart (and the other way around).

And kudos for adding keyboard navigation! Everything looks good to me.

mikejolley commented:
> Yes! Do we have some logic in place to take the user to one or the other if either cart or checkout is not present? I want to make sure we're accommodating use cases where the user only has block checkout, but not cart (and the other way around).

@jarekmorawski Right now its is setup to default to cart if the cart page contains either a cart block, or the classic cart placeholder. If it does not, it will fallback to checkout.

jarekmorawski commented:
Thanks, @mikejolley. I'm testing the flow and though I'm seeing the warning message in the inspector, there's no task on the Home screen. What did I miss? There's one small copy change to make in the inspector notice. I'll ping you in Figma.

mikejolley commented:
@jarekmorawski The task is coded to only show up if you've got a cart/checkout page setup and that page contains the blocks. 

So under pages, it should look like this:
![Screenshot 2023-10-19 at 12 09 17](https://github.com/woocommerce/woocommerce-blocks/assets/90977/aba6b5e9-35b4-4cb3-9ff1-4447b2fc5d03)

And then those pages need to contain the blockified cart/checkout.

jarekmorawski commented:
Wow! Thank you, I've learned something new today. I couldn't see the task because I went straight to the editor and added the blocks directly to the cart and checkout templates. Is there a reason why the default pages don't use them?

mikejolley commented:
@jarekmorawski The defaults will change in the next release, its being worked on separately!

mikejolley committed:
Update modal content

mikejolley committed:
Update modal usage

mikejolley committed:
Removed undo link when converting from classic shortcode

senadir commented on assets/js/blocks/cart/edit.js:
```diff
@@ -49,6 +52,19 @@ export const Edit = ( { className, attributes, setAttributes } ) => {
+				'edit-site/block-inspector'
+			);
+		}
+	}, [ clientId, focus ] );
```
I left a comment about this in slack, but as I said, certain transforms (group, paid content) would change the client id temporarily on hover, which breaks this code.

pmcpinto commented:
> Untested at present but I added 3 events in https://github.com/woocommerce/woocommerce-blocks/commit/659926a28de9457b335072b3aaccdfbffbf7ad43:

@mikejolley  thanks! I'm assuming that each event will have a prop to identify the block (cart or checkout), right?

senadir reviewed (approved):
Additional changes make sense, left a single comment about an edge case but once that's fixed we can merge this.

mikejolley committed:
Check if block was selected

mikejolley committed:
Revert "Removed undo link when converting from classic shortcode"

This reverts commit 2babbab4c1e69861a0371ff745e85d80ff6bbab1.

mikejolley committed:
update snackbar text

ralucaStan reviewed (approved):
Great work here Mike. Thank you for your flexibility on this one and for taking the time to accommodate the tracking changes. 👏

senadir commented:
@pmcpinto yes each event will have a prop named shortcode be cart or checkout

pmcpinto commented:
> @pmcpinto yes each event will have a prop named shortcode be cart or checkout

Thanks for the confirmation
